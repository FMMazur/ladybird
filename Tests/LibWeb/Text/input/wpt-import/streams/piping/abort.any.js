// META: global=window,worker,shadowrealm
// META: script=../resources/recording-streams.js
// META: script=../resources/test-utils.js
'use strict';

// Tests for the use of pipeTo with AbortSignal.
// There is some extra complexity to avoid timeouts in environments where abort is not implemented.

const error1 = new Error('error1');
error1.name = 'error1';
const error2 = new Error('error2');
error2.name = 'error2';

const errorOnPull = {
  pull(controller) {
    // This will cause the test to error if pipeTo abort is not implemented.
    controller.error('failed to abort');
  }
};

// To stop pull() being called immediately when the stream is created, we need to set highWaterMark to 0.
const hwm0 = { highWaterMark: 0 };

promise_test(t => {
  const rs = recordingReadableStream(errorOnPull, hwm0);
  const ws = new WritableStream();
  const abortController = new AbortController();
  const signal = abortController.signal;
  abortController.abort();
  return promise_rejects_dom(t, 'AbortError', rs.pipeTo(ws, { signal }), 'pipeTo should reject')
      .then(() => Promise.all([
        rs.getReader().closed,
        promise_rejects_dom(t, 'AbortError', ws.getWriter().closed, 'writer.closed should reject')
      ]))
      .then(() => {
        assert_equals(rs.events.length, 2, 'cancel should have been called');
        assert_equals(rs.events[0], 'cancel', 'first event should be cancel');
        assert_equals(rs.events[1].name, 'AbortError', 'the argument to cancel should be an AbortError');
        assert_equals(rs.events[1].constructor.name, 'DOMException',
                      'the argument to cancel should be a DOMException');
      });
}, 'an aborted signal should cause the writable stream to reject with an AbortError');
