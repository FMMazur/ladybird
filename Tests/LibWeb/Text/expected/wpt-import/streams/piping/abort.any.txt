Harness status: OK

Found 6 tests

5 Pass
1 Fail
Pass	a signal argument 'null' should cause pipeTo() to reject
Pass	a signal argument 'AbortSignal' should cause pipeTo() to reject
Pass	a signal argument 'true' should cause pipeTo() to reject
Pass	a signal argument '-1' should cause pipeTo() to reject
Pass	a signal argument '[object AbortSignal]' should cause pipeTo() to reject
Fail	an aborted signal should cause the writable stream to reject with an AbortError